gotham/src/os/windows.rs

```diff
@@ -1,152 +1,108 @@
-    queue: SocketQueue,
-    handle: &Handle,
-) -> Box<Future<Item = (), Error = io::Error>> {
-    let listener = tokio_core::net::TcpListener::from_listener(listener, &addr, handle)
-        .expect("unable to convert TCP listener to tokio listener");
-
-    let mut n: usize = 0;
-
-    Box::new(listener.incoming().for_each(move |conn| {
-        queue.queue.push(conn);
-        let tasks = queue
-            .notify
-            .lock()
-            .expect("mutex poisoned, futures::task::Task::notify panicked?");
-
-        n = (n + 1) % tasks.len();
-        tasks[n].notify();
-        Ok(())
-    }))
-}
-
-fn start_serve_core<NH>(queue: SocketQueue, protocol: &Http, new_handler: Arc<NH>)
-where
-    NH: NewHandler + 'static,
-{
-    let mut core = Core::new().expect("unable to spawn tokio reactor");
-    let handle = core.handle();
-    core.run(serve(queue, protocol, new_handler, &handle))
-        .expect("unable to run reactor for work stealing");
-}
+    fn incoming(self, handle: Handle) -> Self::Stream {
+        let tasks_m = self.notify.clone();
 
-fn serve<'a, NH>(
-    queue: SocketQueue,
-    protocol: &'a Http,
-    new_handler: Arc<NH>,
-    handle: &'a Handle,
-) -> Box<Future<Item = (), Error = ()> + 'a>
-where
-    NH: NewHandler + 'static,
-{
-    let gotham_service = GothamService::new(new_handler, handle.clone());
-    let tasks_m = queue.notify.clone();
-
-    Box::new(
-        future::lazy(move || {
+        handle.spawn(future::lazy(move || {
             let mut tasks = tasks_m
                 .lock()
                 .expect("mutex poisoned, futures::task::Task::notify panicked?");
             tasks.push(task::current());
             future::ok(())
-        }).and_then(move |_| {
-            queue.for_each(move |(socket, addr)| {
-                let service = gotham_service.connect(addr);
-                let f = protocol.serve_connection(socket, service).then(|_| Ok(()));
-
-                handle.spawn(f);
-                Ok(())
-            })
-        }),
-    )
+        }));
+
+        SocketStream {
+            queue: self.queue.clone(),
+        }
+    }
+}
+
+pub struct SocketStream {
+    queue: Arc<SegQueue<(TcpStream, SocketAddr)>>,
+}
+
+impl Stream for SocketStream {
+    type Item = (TcpStream, SocketAddr);
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.queue.try_pop() {
+            Some(t) => Ok(Async::Ready(Some(t))),
+            None => Ok(Async::NotReady),
+        }
+    }
+}
+
+/// Constructs a GothamListener to handle incoming TCP connections.
+///
+/// Note: On Windows this function spawns an extra thread to handle
+/// accepting connections.
+pub fn new_gotham_listener(addr: SocketAddr) -> SocketQueue {
+    let tcp = TcpListener::bind(addr).expect("unable to open TCP listener");
+    let queue = SocketQueue::new(tcp, addr);
+    {
+        let queue = queue.clone();
+        thread::spawn(move || queue.listen());
```

Comment:
Is it fair to say that `tcp` and `addr` could be arguments to this `queue.listen()` call rather than being stored in the `SocketQueue` struct? Having an `Arc<TcpListener>` there seems to give the wrong impression at a casual read – that the listener is thread-safe – when it is not. Refactoring this might help prevent confusion. Thoughts?

Reply:
> The `SocketQueue` winds up collecting several values, and my first instinct is to gather them together in the constructor.

I can appreciate that instinct. I was coming at it from the other direction, where I know they're only used in a single place and by a single thread (out of necessity, because the listener can't be shared).

I note that your current revision doesn't build due to an issue with the import. I thought that an `Rc` might cause problems here by making the `SocketQueue` struct not `Send + Sync`, though I could be wrong about that.

If the `Rc` version works, it does the "right thing" by preventing the listener being shared.